test/hello/test/src/test_utils.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:io';
+
+import 'package:test/test.dart';
+import 'package:test_process/test_process.dart';
+
+const defaultPort = 8080;
+
+int _autoPort;
+
+int get autoPort => _autoPort;
+
+Future<TestProcess> startServerTest({
+  bool shouldFail = false,
+  int expectedListeningPort = defaultPort,
+  Map<String, String> env,
+  Iterable<String> arguments = const <String>[],
+}) async {
+  if (expectedListeningPort == 0) {
+    expect(arguments, isNot(contains('--port')));
+    if (env != null) {
+      expect(env, isNot(contains('PORT')));
+    }
+  }
+  final args = [
+    'bin/server.dart',
+    ...arguments,
+    if (expectedListeningPort == 0) ...['--port', '0'],
+  ];
+  final proc = await TestProcess.start('dart', args, environment: env);
+
+  if (!shouldFail) {
+    final output = await proc.stdout.next;
+    final match = _listeningPattern.firstMatch(output);
+    expect(match, isNotNull);
+    _autoPort = int.parse(match[1]);
+    if (expectedListeningPort == 0) {
```

Comment:
If dynamically assigning ports for parallel testing, might want restrict them to the dynamic port range: 49152 to 65535. Well-known ports are 0-1023, registered ports are in the range 1024 to 49151.

Reply:
Dart's logic here: if you give us `0` we will use a port in the dynamic range. Make sense?